src/main/java/org/alliancegenome/curation_api/services/validation/AnnotationValidator.java

```diff
@@ -0,0 +1,228 @@
+import org.alliancegenome.curation_api.dao.OrganizationDAO;
+import org.alliancegenome.curation_api.dao.ReferenceDAO;
+import org.alliancegenome.curation_api.model.entities.Annotation;
+import org.alliancegenome.curation_api.model.entities.ConditionRelation;
+import org.alliancegenome.curation_api.model.entities.DataProvider;
+import org.alliancegenome.curation_api.model.entities.Note;
+import org.alliancegenome.curation_api.model.entities.Reference;
+import org.alliancegenome.curation_api.response.ObjectResponse;
+import org.alliancegenome.curation_api.response.SearchResponse;
+import org.alliancegenome.curation_api.services.DataProviderService;
+import org.alliancegenome.curation_api.services.ReferenceService;
+import org.alliancegenome.curation_api.services.VocabularyTermService;
+import org.alliancegenome.curation_api.services.helpers.notes.NoteIdentityHelper;
+import org.apache.commons.collections.CollectionUtils;
+import org.apache.commons.collections.ListUtils;
+import org.apache.commons.lang3.ObjectUtils;
+import org.apache.commons.lang3.StringUtils;
+
+public class AnnotationValidator extends AuditedObjectValidator<Annotation> {
+
+	@Inject
+	ReferenceValidator referenceValidator;
+	@Inject
+	NoteValidator noteValidator;
+	@Inject
+	NoteDAO noteDAO;
+	@Inject
+	ConditionRelationValidator conditionRelationValidator;
+	@Inject
+	ConditionRelationDAO conditionRelationDAO;
+	@Inject
+	AnnotationDAO annotationDAO;
+	@Inject
+	DataProviderService dataProviderService;
+	@Inject
+	DataProviderValidator dataProviderValidator;
+	@Inject
+	DataProviderDAO dataProviderDAO;
+
+		
+	public DataProvider validateDataProvider(Annotation uiEntity, Annotation dbEntity) {
+		String field = "dataProvider";
+		
+		if (uiEntity.getDataProvider() == null) {
+			if (dbEntity.getId() == null)
+				uiEntity.setDataProvider(dataProviderService.createAffiliatedModDataProvider());
+			if (uiEntity.getDataProvider() == null) {
+				addMessageResponse(field, ValidationConstants.REQUIRED_MESSAGE);
+				return null;
+			}
+		} 
+
+		DataProvider uiDataProvider = uiEntity.getDataProvider();
+		DataProvider dbDataProvider = dbEntity.getDataProvider();
+		
+		ObjectResponse<DataProvider> dpResponse = dataProviderValidator.validateDataProvider(uiDataProvider, dbDataProvider, false);
+		if (dpResponse.hasErrors()) {
+			addMessageResponse(field, dpResponse.errorMessagesString());
+			return null;
+		}
+		
+		DataProvider validatedDataProvider = dpResponse.getEntity();
+		if (validatedDataProvider.getObsolete() && (dbDataProvider == null || !validatedDataProvider.getId().equals(dbDataProvider.getId()))) {
+			addMessageResponse(field, ValidationConstants.OBSOLETE_MESSAGE);
+			return null;
+		}
+		
+		if (validatedDataProvider.getId() == null)
+			validatedDataProvider = dataProviderDAO.persist(validatedDataProvider);
+		
+		return validatedDataProvider;
+	}
+
+	public List<Note> validateRelatedNotes(Annotation uiEntity, Annotation dbEntity) {
```

Comment:
Is this note validation code somewhere else also?

Reply:
The validation of the note object itself is handled by NoteValidator.java.  This method does some extra validation specific to annotations and some cleanup.  There is a similar, but not identical, method in AlleleValidator.java.

Reply:
Ahh ok